deku-derive/src/macros/mod.rs

```diff
@@ -65,3 +59,147 @@ fn gen_struct_init<I: ToTokens>(is_named: bool, field_idents: Vec<I>) -> TokenSt
+
+    syn::Ident::new(&name, span).to_token_stream()
+}
+
+/// -> `{ a: __a }` or `(__a)`
+fn gen_internal_field_idents(named: bool, idents: Vec<TokenStream>) -> Vec<TokenStream> {
+    // -> `{ a: __a }` or `(__a)`
+    if named {
+        idents
+            .into_iter()
+            .map(|i| (i.clone(), gen_internal_field_ident(i)))
+            .map(|(i, h)| quote! {#i: #h})
+            .collect()
+    } else {
+        idents.into_iter().map(gen_internal_field_ident).collect()
+    }
+}
+
+fn split_ctx_to_pats_and_types(
+    ctx: &Punctuated<syn::FnArg, syn::token::Comma>,
+) -> syn::Result<Vec<(&syn::Pat, &syn::Type)>> {
+    // `()` or `(u8, u32)`
+    ctx.iter()
+        .map(|arg| {
+            match arg {
+                syn::FnArg::Typed(pat_type) => Ok((pat_type.pat.as_ref(), pat_type.ty.as_ref())),
+                // a self is unacceptable
+                syn::FnArg::Receiver(r) => Err(syn::Error::new(r.span(), "Unacceptable context")),
+            }
+        })
+        .collect::<Result<Vec<_>, _>>()
+}
+
+fn gen_ctx_types_and_arg(
+    ctx: Option<&Punctuated<syn::FnArg, syn::token::Comma>>,
+) -> syn::Result<(TokenStream, TokenStream)> {
+    if let Some(ctx) = ctx {
+        let pats_types = split_ctx_to_pats_and_types(ctx)?;
+
+        if pats_types.len() == 1 {
+            // remove paren for single item
+            let (pat, ty) = pats_types[0];
+            Ok((quote! {#ty}, quote! {#pat:#ty}))
+        } else {
+            let pats = pats_types.iter().map(|(pat, _)| pat);
+            let types = pats_types.iter().map(|(_, ty)| ty);
+
+            // avoid move
+            let types2 = types.clone();
+
+            // "a: u8, b: usize" -> (u8, usize)
+            let ctx_types = quote! {(#(#types2),*)};
+            // "a: u8, b: usize" -> (a, b): (u8, usize)
+            let ctx_arg = quote! {(#(#pats),*): (#(#types),*)};
+
+            Ok((ctx_types, ctx_arg))
+        }
+    } else {
+        Ok((quote! {()}, quote! {_: ()}))
+    }
+}
+
+fn gen_id_args(endian: Option<&syn::LitStr>, bits: Option<usize>) -> TokenStream {
+    let endian = endian.map(|s| {
+        quote! {<deku::ctx::Endian as core::str::FromStr>::from_str(#s)}
+    });
+
+    let bits = bits.map(|n| quote! {deku::ctx::BitSize(#n)});
+
+    let mut id_args = Vec::with_capacity(2);
+    if let Some(id_endian) = endian {
+        id_args.push(id_endian);
+    }
+    if let Some(id_bits) = bits {
+        id_args.push(id_bits);
+    }
+    if id_args.len() == 1 {
+        let arg = &id_args[0];
+        quote! {#arg}
+    } else {
+        quote! {#(#id_args),*}
+    }
+}
+
+fn gen_field_args(
+    endian: Option<&syn::LitStr>,
+    bits: Option<usize>,
+    ctx: Option<&Punctuated<syn::Expr, syn::token::Comma>>,
+) -> syn::Result<TokenStream> {
+    let mut args = Vec::with_capacity(3);
+
+    if let Some(field_endian) = endian {
+        args.push(match field_endian.value().as_str() {
+            "little" => quote! {deku::ctx::Endian::Little},
```

Comment:
`from_str` + map_err here?

Reply:
At first I think the same way with you. But when I try to do it, I found I must add `deku` as a dependency.

Reply:
Ah! Makes sense.